design-documents/graph-ql/mutation-error-design.md

```diff
@@ -0,0 +1,153 @@
+
+## Example of the problem
+
+Take this simplified schema to add an item to a shopper's cart:
+
+```graphql
+type Mutation {
+    addItemToCart(sku: String!, quantity: Float): Cart
+}
+```
+
+This schema only describes one side-effect that can happen as a result of adding an item to the cart: it succeeds! But some things can go wrong when adding an item to the cart, and we know what many of those things are:
+
+- Item went out of stock since the product page was loaded
+- Merchant disabled the product since the product page was loaded
+- Cart hit the max quantity allowed of item per customer. Full qty selected was not added, but some were
+- Extensions can add additional rules that would limit adding an item to the cart
+
+Now put yourself in the shoes of a UI developer: how do you know all the error states your UI needs to cover? For most of the Magento 2 mutations today, you would need run the mutation itself (or dig through PHP) to see what will be returned in the `errors` key of the GraphQL response:
+
+```js
+{
+    "data": {
+        "cart": {
+            // cart data here
+        }
+    },
+    "errors": [{
+        "message": "Item 'cool-hat' not added to cart (Out of Stock)",
+        "path": ["addItemToCart"]
+    }]
+}
+```
+
+This has some problems:
+
+1. It's unreasonable to expect a UI developer to exercise every possible input to a resolver to find the error states manually
+2. These errors are not enforced by the schema, so they're not versioned with the schema.
+3. If the UI wants to customize the message for a specific error state, they'd have to match on the exact response string, because there's no concept of error codes
+
+Until recently, we didn't need to think quite as carefully about modeling errors and versioning them, because we owned the single reference theme for the application (Luma). In the world of headless UIs, it's critical that our errors are versioned and discoverable.
+
+## Solution: Design Mutation errors into the schema
+
+If we move our known error states for various mutations into the schema design itself, we'll get a few benefits:
+
+1. Versioning of errors
+2. Discoverability of error states for UI devs
+
+### Example: Fixing our `addItemToCart` mutation
+
+In the previous example of `addItemToCart`, we had the following schema:
+
+```graphql
+type Mutation {
+    addItemToCart(sku: String!, quantity: Float): Cart
+}
+```
+
+Let's design some known error states directly into the API
+
+```graphql
+type Mutation {
+    addItemToCart(sku: String!, quantity: Float): AddItemToCartOutput
+}
+
+type AddItemToCartOutput {
+    # can still get the entire state of the cart post-mutation
+    cart: Cart
+    # can directly access errors that a shopper should be notified about
+    addItemUserErrors: [AddItemUserError!]!
+}
+
+type AddItemUserError {
+    # if a UI has their own text for message, they can just
```

Comment:
Let's add `path` and `location`, see https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/customer-reorder.graphqls

Reply:
@DrewML I see `path` in the updated schema, what about `location`?

Reply:
@DrewML This looks good https://github.com/magento/architecture/pull/393/files#r453972900. It makes sense to standardize this approach. For alignment with existing reorder mutation and convention, can we use 'code' instead of 'type'. Also another suggestion came up Alex when we discussed this, field names should be snake cased. So suggesting 'user_errors' instead of 'userErrors'.